### PR TITLE
Change focal length in sfmData file formats

### PR DESCRIPTION
**Pull request by fabiencastan**
## Description

Changes storage of focal in file formats:

- [x] Focal is now in millimeter (instead of pixels).
- [x] Focal is no more split on (x, y) but use focal and pixel ratio.

### CODE REVIEW

File: src/aliceVision/sfmDataIO/jsonIO.cpp [hunk 1]
```diff
@@ -108,7 +108,13 @@ void saveIntrinsic(const std::string& name, IndexT intrinsicId, const std::share
   if (intrinsicScaleOffset)
   {
     intrinsicTree.put("pxInitialFocalLength", intrinsicScaleOffset->initialScale());
```
Comment (fabiencastan) [inline src/aliceVision/sfmDataIO/jsonIO.cpp:110]: pxInitialFocalLength needs to be fixed.
And also in ABC

File: src/aliceVision/sfmDataIO/jsonIO.cpp [hunk 2]
```diff
@@ -108,7 +108,13 @@ void saveIntrinsic(const std::string& name, IndexT intrinsicId, const std::share
...
-    saveMatrix("pxFocalLength", intrinsicScaleOffset->getScale(), intrinsicTree);
+    
+    double focalLengthMM = intrinsicScaleOffset->sensorWidth() * intrinsicScaleOffset->getScale().x() / double(intrinsic->w());
+    double pixelRatio = (intrinsicScaleOffset->getScale().x()) / intrinsicScaleOffset->getScale().y();
+
+    intrinsicTree.put("pxFocalLength", focalLengthMM);
```
Comment (fabiencastan) [inline src/aliceVision/sfmDataIO/jsonIO.cpp:115]: Need to change name

File: src/aliceVision/sfmDataIO/jsonIO.cpp [hunk 3]
```diff
@@ -108,7 +108,13 @@ void saveIntrinsic(const std::string& name, IndexT intrinsicId, const std::share
...
+    intrinsicTree.put("pixelRatio", pixelRatio);
```
Comment (fabiencastan) [inline src/aliceVision/sfmDataIO/jsonIO.cpp:116]: pixelRatio needs to be in ABC IO too

File: src/aliceVision/sfmDataIO/AlembicImporter.cpp [hunk 4]
```diff
@@ -427,6 +428,10 @@ bool readCamera(const Version & abcVersion, const ICamera& camera, const M44d& m
       {
         intrinsicLocked = getAbcProp<Alembic::Abc::IBoolProperty>(userProps, *propHeader, "mvg_intrinsicLocked", sampleFrame);
       }
+      if(const Alembic::Abc::PropertyHeader *propHeader = userProps.getPropertyHeader("mvg_intrinsicFocalRatioLocked"))
+      {
+        lockRatio = getAbcProp<Alembic::Abc::IBoolProperty>(userProps, *propHeader, "mvg_intrinsicFocalRatioLocked", sampleFrame);
+      }
```
Comment (fabiencastan) [inline src/aliceVision/sfmDataIO/AlembicImporter.cpp:434]: To be symetric with json
```suggestion
      if(const Alembic::Abc::PropertyHeader *propHeader = userProps.getPropertyHeader("mvg_intrinsicPixelRatioLock"))
      {
        lockRatio = getAbcProp<Alembic::Abc::IBoolProperty>(userProps, *propHeader, "mvg_intrinsicPixelRatioLock", sampleFrame);
      }
```

**Comment by simogasp** [comment]
it would be better to have the possibility to keep the focal length in pixel as it is the standard way in computer vision to define the intrinsics. without it will be hard to e.g. run evaluation benchmarks that probably won't have the camera information to recover the information in mm

File: src/aliceVision/sfmDataIO/viewIO.cpp [hunk 5]
```diff
@@ -108,41 +108,43 @@ void updateIncompleteView(sfmData::View& view, EViewIdMethod viewIdMethod, const
...
   // get view informations
   const std::string& cameraBrand = view.getMetadataMake();
   const std::string& cameraModel = view.getMetadataModel();
   const std::string& bodySerialNumber = view.getMetadataBodySerialNumber();
   const std::string& lensSerialNumber = view.getMetadataLensSerialNumber();
 
-  double focalLengthIn35mm = mmFocalLength; // crop factor is apply later if sensor width is defined
-
-  double pxFocalLength;
+  double focalLength;
```
Comment (fabiencastan) [inline src/aliceVision/sfmDataIO/viewIO.cpp:125]:
```suggestion
  double focalLength{-1.0};
```

File: src/aliceVision/sfmDataIO/jsonIO.cpp [hunk 6]
```diff
@@ -169,10 +177,21 @@ void loadIntrinsic(const Version & version, IndexT& intrinsicId, std::shared_ptr
...
   {
     loadMatrix("pxFocalLength", pxFocalLength, intrinsicTree);
   }
+  else 
+  {
+    double fmm = intrinsicTree.get<double>("focalLength", 1.0);
+    double ratio = intrinsicTree.get<double>("pixelRatio", 1.0);
+
+    double px = (fmm / sensorWidth) * double(width);
+    double py = px / ratio;
```
Comment (fabiencastan) [inline src/aliceVision/sfmDataIO/jsonIO.cpp:190]:
```suggestion
    const double fmm = intrinsicTree.get<double>("focalLength", 1.0);
    const double ratio = intrinsicTree.get<double>("pixelRatio", 1.0);

    const double px = (fmm / sensorWidth) * double(width);
    const double py = px / ratio;
```

File: src/aliceVision/sfmDataIO/jsonIO.cpp [hunk 7]
```diff
@@ -107,8 +107,16 @@ void saveIntrinsic(const std::string& name, IndexT intrinsicId, const std::share
   std::shared_ptr<camera::IntrinsicsScaleOffset> intrinsicScaleOffset = std::dynamic_pointer_cast<camera::IntrinsicsScaleOffset>(intrinsic);
   if (intrinsicScaleOffset)
   {
-    intrinsicTree.put("pxInitialFocalLength", intrinsicScaleOffset->initialScale());
-    saveMatrix("pxFocalLength", intrinsicScaleOffset->getScale(), intrinsicTree);
+    
+    double initialFocalLengthMM = intrinsicScaleOffset->sensorWidth() * intrinsicScaleOffset->getInitialScale().x() / double(intrinsic->w());
+    double focalLengthMM = intrinsicScaleOffset->sensorWidth() * intrinsicScaleOffset->getScale().x() / double(intrinsic->w());
+    double pixelRatio = (intrinsicScaleOffset->getScale().x()) / intrinsicScaleOffset->getScale().y();
```
Comment (fabiencastan) [inline src/aliceVision/sfmDataIO/jsonIO.cpp:113]:
```suggestion
    const double initialFocalLengthMM = intrinsicScaleOffset->sensorWidth() * intrinsicScaleOffset->getInitialScale().x() / double(intrinsic->w());
    const double focalLengthMM = intrinsicScaleOffset->sensorWidth() * intrinsicScaleOffset->getScale().x() / double(intrinsic->w());
    const double pixelRatio = (intrinsicScaleOffset->getScale().x()) / intrinsicScaleOffset->getScale().y();
```

**Pull request merged**